File: src/main/java/com/company/erp/features/employees/service/EmployeeAvatarService.java
package com.company.erp.features.employees.service;

import com.company.erp.core.config.AppProperties;
import com.company.erp.core.exceptions.BadRequestException;
import com.company.erp.features.employees.entity.Employee;
import com.company.erp.features.employees.repository.EmployeeRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.time.Instant;
import java.util.Arrays;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@Transactional
public class EmployeeAvatarService {

    private final EmployeeService employees;
    private final EmployeeRepository employeeRepo;
    private final Path avatarDir;
    private final String publicBaseUrl;
    private final long maxBytes;
    private final Set<String> allowedTypes;

    public EmployeeAvatarService(EmployeeService employees,
                                 EmployeeRepository employeeRepo,
                                 AppProperties props) {
        this.employees = employees;
        this.employeeRepo = employeeRepo;
        AppProperties.Uploads.Avatar cfg = props.uploads().avatar();
        this.avatarDir     = Paths.get(cfg.dir()).toAbsolutePath().normalize();
        this.publicBaseUrl = trimTrailingSlash(cfg.publicBaseUrl());
        this.maxBytes      = cfg.maxFileSize();
        this.allowedTypes  = Arrays.stream(cfg.allowedContentTypes().split(","))
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .collect(Collectors.toUnmodifiableSet());
    }

    public Employee uploadForEmployee(Long employeeId, MultipartFile file) {
        Employee employee = employees.getById(employeeId);
        return store(employee, file);
    }

    public Employee uploadForCurrentUser(Long userId, MultipartFile file) {
        Employee employee = employees.getByUserId(userId);
        return store(employee, file);
    }

    public Employee deleteForEmployee(Long employeeId) {
        return clear(employees.getById(employeeId));
    }

    public Employee deleteForCurrentUser(Long userId) {
        return clear(employees.getByUserId(userId));
    }

    private Employee store(Employee employee, MultipartFile file) {
        if (file == null || file.isEmpty()) {
            throw new BadRequestException("Avatar file is required");
        }
        if (file.getSize() > maxBytes) {
            throw new BadRequestException("Avatar exceeds max size of " + maxBytes + " bytes");
        }
        String contentType = file.getContentType();
        if (contentType == null || !allowedTypes.contains(contentType)) {
            throw new BadRequestException("Unsupported content type: " + contentType
                    + " (allowed: " + allowedTypes + ")");
        }

        String ext = extensionFor(contentType);
        String filename = employee.getId() + "-" + UUID.randomUUID() + "." + ext;
        Path target = avatarDir.resolve(filename);

        try {
            Files.createDirectories(avatarDir);
            try (var in = file.getInputStream()) {
                Files.copy(in, target, StandardCopyOption.REPLACE_EXISTING);
            }
        } catch (IOException ex) {
            throw new RuntimeException("Failed to store avatar: " + ex.getMessage(), ex);
        }

        deleteFileQuietly(employee.getAvatarUrl());
        employee.setAvatarUrl(publicBaseUrl + "/" + filename);
        employee.setAvatarContentType(contentType);
        employee.setAvatarUploadedAt(Instant.now());
        return employeeRepo.save(employee);
    }

    private Employee clear(Employee employee) {
        if (employee.getAvatarUrl() != null) {
            deleteFileQuietly(employee.getAvatarUrl());
            employee.setAvatarUrl(null);
            employee.setAvatarContentType(null);
            employee.setAvatarUploadedAt(null);
        }
        return employee;
    }

    private void deleteFileQuietly(String publicUrl) {
        if (publicUrl == null || !publicUrl.startsWith(publicBaseUrl + "/")) return;
        String filename = publicUrl.substring((publicBaseUrl + "/").length());
        if (filename.contains("/") || filename.contains("\\") || filename.contains("..")) return;
        try {
            Files.deleteIfExists(avatarDir.resolve(filename));
        } catch (IOException ignored) {
            // best-effort cleanup; missing files are not fatal
        }
    }

    private static String extensionFor(String contentType) {
        return switch (contentType) {
            case "image/jpeg" -> "jpg";
            case "image/png"  -> "png";
            case "image/webp" -> "webp";
            default           -> "bin";
        };
    }

    private static String trimTrailingSlash(String s) {
        return s.endsWith("/") ? s.substring(0, s.length() - 1) : s;
    }
}

File: app/Features/Employees/Services/EmployeeAvatarService.php
<?php

namespace App\Features\Employees\Services;

use App\Features\Employees\Models\Employee;
use App\Support\Exceptions\BadRequestException;
use Illuminate\Http\UploadedFile;
use Illuminate\Support\Carbon;
use Illuminate\Support\Str;

/**
 * Server-side multipart avatar upload — files are stored on the API host's
 * filesystem and served back as static resources. Port of the Spring
 * EmployeeAvatarService.
 */
class EmployeeAvatarService
{
    /** @var EmployeeService */
    private $employees;

    /** @var string */
    private $avatarDir;

    /** @var string */
    private $publicBaseUrl;

    /** @var int */
    private $maxBytes;

    /** @var array<int, string> */
    private $allowedTypes;

    public function __construct(EmployeeService $employees)
    {
        $this->employees = $employees;
        $this->avatarDir = $this->resolveDir((string) config('erp.uploads.avatar.dir'));
        $this->publicBaseUrl = rtrim((string) config('erp.uploads.avatar.public_base_url'), '/');
        $this->maxBytes = (int) config('erp.uploads.avatar.max_file_size');
        $this->allowedTypes = array_values(array_filter(array_map('trim',
            explode(',', (string) config('erp.uploads.avatar.allowed_content_types')))));
    }

    public function uploadForEmployee(int $employeeId, ?UploadedFile $file): Employee
    {
        return $this->store($this->employees->getById($employeeId), $file);
    }

    public function uploadForCurrentUser(int $userId, ?UploadedFile $file): Employee
    {
        return $this->store($this->employees->getByUserId($userId), $file);
    }

    public function deleteForEmployee(int $employeeId): Employee
    {
        return $this->clear($this->employees->getById($employeeId));
    }

    public function deleteForCurrentUser(int $userId): Employee
    {
        return $this->clear($this->employees->getByUserId($userId));
    }

    private function store(Employee $employee, ?UploadedFile $file): Employee
    {
        if ($file === null || ! $file->isValid()) {
            throw new BadRequestException('Avatar file is required');
        }
        if ($file->getSize() > $this->maxBytes) {
            throw new BadRequestException('Avatar exceeds max size of '.$this->maxBytes.' bytes');
        }
        $contentType = $file->getClientMimeType();
        if ($contentType === null || ! in_array($contentType, $this->allowedTypes, true)) {
            throw new BadRequestException('Unsupported content type: '.$contentType
                .' (allowed: ['.implode(', ', $this->allowedTypes).'])');
        }

        $ext = $this->extensionFor($contentType);
        $filename = $employee->id.'-'.Str::uuid().'.'.$ext;

        if (! is_dir($this->avatarDir)) {
            @mkdir($this->avatarDir, 0775, true);
        }
        $file->move($this->avatarDir, $filename);

        $this->deleteFileQuietly($employee->avatar_url);
        $employee->avatar_url = $this->publicBaseUrl.'/'.$filename;
        $employee->avatar_content_type = $contentType;
        $employee->avatar_uploaded_at = Carbon::now();
        $employee->save();

        return $employee;
    }

    private function clear(Employee $employee): Employee
    {
        if ($employee->avatar_url !== null) {
            $this->deleteFileQuietly($employee->avatar_url);
            $employee->avatar_url = null;
            $employee->avatar_content_type = null;
            $employee->avatar_uploaded_at = null;
            $employee->save();
        }
        return $employee;
    }

    private function deleteFileQuietly(?string $publicUrl): void
    {
        if ($publicUrl === null || strpos($publicUrl, $this->publicBaseUrl.'/') !== 0) {
            return;
        }
        $filename = substr($publicUrl, strlen($this->publicBaseUrl.'/'));
        if (strpos($filename, '/') !== false || strpos($filename, '\\') !== false || strpos($filename, '..') !== false) {
            return;
        }
        $path = $this->avatarDir.DIRECTORY_SEPARATOR.$filename;
        if (is_file($path)) {
            @unlink($path);
        }
    }

    private function extensionFor(string $contentType): string
    {
        switch ($contentType) {
            case 'image/jpeg':
                return 'jpg';
            case 'image/png':
                return 'png';
            case 'image/webp':
                return 'webp';
            default:
                return 'bin';
        }
    }

    private function resolveDir(string $dir): string
    {
        if ($dir === '') {
            $dir = base_path('uploads/avatars');
        }
        // Absolute path stays; relative is resolved against the project root.
        if ($dir[0] !== '/' && ! preg_match('/^[A-Za-z]:[\\\\\\/]/', $dir)) {
            $dir = base_path(ltrim(preg_replace('#^\./#', '', $dir), '/'));
        }
        return rtrim($dir, '/\\');
    }
}
